pkgs/shells/fish/default.nix

```diff
@@ -35,25 +35,55 @@ let
     #     source both, but source the more global configuration files earlier
     #     than the more local ones, so that more local configurations inherit
     #     from but override the more global locations.
-    
+
     if test -f /etc/fish/config.fish
       source /etc/fish/config.fish
     end
-    
+
     #                                                                             #
     ############### ↑ Nix hook for sourcing /etc/fish/config.fish ↑ ###############
   '';
 
   fishPreInitHooks = ''
-    # source nixos environment
+    # source nix environment
     # note that this is required:
     #   1. For all shells, not just login shells (mosh needs this as do some other command-line utilities)
     #   2. Before the shell is initialized, so that config snippets can find the commands they use on the PATH
-    builtin status --is-login
-    or test -z "$__fish_nixos_env_preinit_sourced" -a -z "$ETC_PROFILE_SOURCED" -a -z "$ETC_ZSHENV_SOURCED"
-    and test -f /etc/fish/nixos-env-preinit.fish
-    and source /etc/fish/nixos-env-preinit.fish
-    and set -gx __fish_nixos_env_preinit_sourced 1
+    if builtin status --is-login
+       or test -z "$__fish_nix_env_preinit_sourced" -a -z "$ETC_PROFILE_SOURCED" -a -z "$ETC_ZSHENV_SOURCED"
+
+      set -l nix_env_setup
+      # If we have a configuration provided by a NixOS or Nix-Darwin administrator, prefer it.
+      if test -f /etc/nix/support/set-environment.sh
+        set nix_env_setup /etc/nix/support/set-environment.sh
+      else
+        set -l common_prefix @NIX_STORE/../var/nix/profiles/default/etc/profile.d
+        set -l script_selection nix.sh
+a
```

Comment:
Oops!

```
/nix/store/6f429zrnlkpc8iymsci18w7dgh5km1vl-fish-2.6.0/share/fish/__nix_preinit_env.fish (line 15):
a
^
from sourcing file /nix/store/6f429zrnlkpc8iymsci18w7dgh5km1vl-fish-2.6.0/share/fish/__nix_preinit_env.fish
	called on line 9 of file /nix/store/6f429zrnlkpc8iymsci18w7dgh5km1vl-fish-2.6.0/share/fish/__fish_build_paths.fish

from sourcing file /nix/store/6f429zrnlkpc8iymsci18w7dgh5km1vl-fish-2.6.0/share/fish/__fish_build_paths.fish
	called on line 87 of file /nix/store/6f429zrnlkpc8iymsci18w7dgh5km1vl-fish-2.6.0/share/fish/config.fish

from sourcing file /nix/store/6f429zrnlkpc8iymsci18w7dgh5km1vl-fish-2.6.0/share/fish/config.fish
	called during startup

/nix/store/ab2iwb8ks5day43sw0580hdf54nfa4pv-bash-4.4-p12/bin/bash: @NIX_STORE/../var/nix/profiles/default/etc/profile.d/nix-daemon.sh: No such file or directory
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
```

Reply:
With the following patch applied:

```diff
diff --git a/pkgs/shells/fish/default.nix b/pkgs/shells/fish/default.nix
index 39235f96d6..55b6d67af7 100644
--- a/pkgs/shells/fish/default.nix
+++ b/pkgs/shells/fish/default.nix
@@ -59,7 +59,7 @@ let
       else
         set -l common_prefix @NIX_STORE/../var/nix/profiles/default/etc/profile.d
         set -l script_selection nix.sh
-a
+
         string match -q (${coreutils}/bin/uname) Darwin
         # Use `set -U nix_on_darwin_no_daemon` to tell Fish to use the single-user mode script instead
         and not set -q nix_on_darwin_no_daemon
```

... I get:
> /nix/store/ab2iwb8ks5day43sw0580hdf54nfa4pv-bash-4.4-p12/bin/bash: @NIX_STORE/../var/nix/profiles/default/etc/profile.d/nix-daemon.sh: No such file or directory

Reply:
This patch gets things running smoothly for me:

```diff
diff --git a/pkgs/shells/fish/default.nix b/pkgs/shells/fish/default.nix
index 39235f96d6..30362d7e08 100644
--- a/pkgs/shells/fish/default.nix
+++ b/pkgs/shells/fish/default.nix
@@ -57,9 +57,9 @@ let
       if test -f /etc/nix/support/set-environment.sh
         set nix_env_setup /etc/nix/support/set-environment.sh
       else
-        set -l common_prefix @NIX_STORE/../var/nix/profiles/default/etc/profile.d
+        set -l common_prefix @NIX_STORE@/../var/nix/profiles/default/etc/profile.d
         set -l script_selection nix.sh
-a
+
         string match -q (${coreutils}/bin/uname) Darwin
         # Use `set -U nix_on_darwin_no_daemon` to tell Fish to use the single-user mode script instead
         and not set -q nix_on_darwin_no_daemon
```

Reply:
urgh yeah that's a typo, sorry.

Reply:
Actually it looks like my $PATH is not getting set correctly.

Reply:
I apologize for my sloppiness; I wrote this when I couldn't sleep at 2am last night and I kinda cleaned it up hastily this morning before work. Thanks for helping me test. :-D

Reply:
@yurrriq is that with the `set-environment.sh` file present? I just rebuilt Fish after finding the typos you identified and manually mangled the `/etc/nix/support/set-environment.sh` link so that it would revert to the non-nix-darwin behavior, and everything worked fine.

Everything continues to work when it's using nix-darwin.

Did you remember to rebase the nix-darwin from Colin's PR against current? Those setEnvironment changes need to be in place; the existing state of the PR isn't enough.

Reply:
That part is fixed now.

I'm still having `$PATH` issues, but with these checkouts together, my system works well:
- https://github.com/yurrriq/nixpkgs/commit/267033518b01b69d872139ed428f2170c3118dc0 (https://github.com/NixOS/nixpkgs/pull/30284/commits/8696c0693fd157f6e44a57f0563c16c127fc7b95 rebased on https://github.com/NixOS/nixpkgs/commit/d26f8b5e00b4a436ec8f9b7fb1b55a0dbda440c5)
- https://github.com/nix-hackers/nix-darwin/commit/0ea82db70eb47726f260466a286d6ce9eb9190ff
- https://github.com/yurrriq/dotfiles/commit/f266cfe56062469a6bf0b8bb2faa2ef9a9f5857b